### ISSUE TITLE
Streamline model training pipeline

### ISSUE THREAD

**Issue by abuzarmahmood** [pulakatlab/auto_slide#9]
As a DVC stage

**Comment by abuzarmahmood** [comment]
The issue involves streamlining the model training pipeline in the `pulakatlab/auto_slide` repository by integrating it as a DVC (Data Version Control) stage. The suggestions and changes made by the agents are as follows:

1. **Files Identified for Modification:**
   - **`src/pipeline/model/training.py`**: This file is responsible for the model training logic. It needs modifications to integrate with DVC for tracking experiments and data versioning.
   - **`src/pipeline/run_pipeline.py`**: This file orchestrates the entire pipeline and requires updates to include the new DVC stage for model training.
   - **`src/pipeline/utils.py`**: Contains utility functions used across the pipeline. It should be reviewed to ensure compatibility with DVC integration.

2. **Specific Changes Suggested:**

   - **Integrate DVC into Model Training**
     - **File:** `src/pipeline/model/training.py`
     - **Lines:** 193-199
     - **Description:** Modify the model saving and loss history saving to ensure they are tracked by DVC. This involves saving these artifacts in a DVC-tracked directory.
     - **Code Snippet:**
       ```python
       # Save model
       model_path = os.path.join(artifacts_dir, 'mask_rcnn_model.pth')
       torch.save(model.state_dict(), model_path)
       # DVC: Track the model file
       os.system(f'dvc add {model_path}')
       
       # Save loss histories
       train_loss_path = os.path.join(artifacts_dir, 'train_losses.npy')
       val_loss_path = os.path.join(artifacts_dir, 'val_losses.npy')
       np.save(train_loss_path, all_train_losses)
       np.save(val_loss_path, all_val_losses)
       # DVC: Track the loss files
       os.system(f'dvc add {train_loss_path} {val_loss_path}')
       ```

   - **Update Pipeline Orchestration to Include DVC Stage**
     - **File:** `src/pipeline/run_pipeline.py`
     - **Lines:** 132-134
     - **Description:** Modify the `run_model_training` function to execute the DVC pipeline stage for model training.
     - **Code Snippet:**
       ```python
       if not args.skip_training:
           # Execute the DVC stage for model training
           os.system('dvc repro train_model.dvc')
           if not run_model_training(args.data_dir):
               logger.warning("Model training failed, but continuing pipeline...")
       ```

   - **Ensure Utility Functions Support DVC Integration**
     - **File:** `src/pipeline/utils.py`
     - **Lines:** No specific changes needed based on current content.
     - **Description:** Review utility functions to ensure they do not conflict with DVC operations. No changes are necessary as the current utility functions do not directly interact with DVC.

By implementing these changes, the model training pipeline will be streamlined as a DVC stage, allowing for better experiment tracking and data management.

---
*This response was automatically generated by blech_bot using model gpt-4o*

---
*This response was automatically generated by blech_bot*
